### PR TITLE
GH-37449: [Python] Catch error importing `pyarrow.dataset` in `python/pyarrow/tests/test_substrait.py`

### LINKED ISSUE #37449: [Python] Catch error importing `pyarrow.dataset` in `python/pyarrow/tests/test_substrait.py`
### Describe the bug, including details regarding any error messages, version, and platform.

As noted by @kou in https://github.com/apache/arrow/pull/34834/files#r1306365538, we need to suppress the error when `pyarrow.dataset` cannot be imported.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by ianmcook**
This catches an error importing `pyarrow.dataset` which is causing a CI failure.
* Closes: #37449

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

**Comment by kou** [comment]
https://github.com/ursacomputing/crossbow/actions/runs/6018601141/job/16327072710#step:3:7764

```text
==================================== ERRORS ====================================
_______________ ERROR collecting pyarrow/tests/test_substrait.py _______________
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/runner.py:372: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/python.py:531: in collect
    self._inject_setup_module_fixture()
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/python.py:545: in _inject_setup_module_fixture
    self.obj, ("setUpModule", "setup_module")
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/python.py:310: in obj
    self._obj = obj = self._getobj()
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/python.py:528: in _getobj
    return self._importtestmodule()
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/python.py:617: in _importtestmodule
    mod = import_path(self.path, mode=importmode, root=self.config.rootpath)
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/pathlib.py:565: in import_path
    importlib.import_module(module_name)
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1050: in _gcd_import
    ???
<frozen importlib._bootstrap>:1027: in _find_and_load
    ???
<frozen importlib._bootstrap>:1006: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:688: in _load_unlocked
    ???
/miniconda-for-arrow/envs/pyarrow-3.10/lib/python3.10/site-packages/_pytest/assertion/rewrite.py:178: in exec_module
    exec(co, module.__dict__)
pyarrow/tests/test_substrait.py:936: in <module>
    pc.equal(ds.field("x"), 7),
E   AttributeError: 'NoneType' object has no attribute 'field'
```

It seems that we need to avoid using `ds` here:

https://github.com/apache/arrow/blob/2bb6be1d337583678b0661a8ecfcb2b0f3291a02/python/pyarrow/tests/test_substrait.py#L932-L933

**Comment by ianmcook** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

**Comment by ianmcook** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

File: python/pyarrow/tests/test_substrait.py [hunk 1]
```diff
@@ -928,11 +933,12 @@ def table_provider(names, _):
...
 @pytest.mark.parametrize("expr", [
-    pc.equal(ds.field("x"), 7),
-    pc.equal(ds.field("x"), ds.field("y")),
-    ds.field("x") > 50
+    "pc.equal(ds.field('x'), 7)",
+    "pc.equal(ds.field('x'), ds.field('y'))",
+    "ds.field('x') > 50"
 ])
 def test_serializing_expressions(expr):
+    expr = eval(expr)
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_substrait.py:941]:
```suggestion
    pc.equal(pc.field('x'), 7)",
    pc.equal(pc.field('x'), pc.field('y'))",
    pc.field('x') > 50"
])
def test_serializing_expressions(expr):
```

The `field` nowadays lives in `pyarrow.compute` (it's only still exposed through `pyarrow.dataset` as well for back compat), but so can be imported from there and is always available at the moment.

**Comment by ianmcook** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

File: python/pyarrow/tests/test_substrait.py (hunk 1 above)
Comment (ianmcook) [reply python/pyarrow/tests/test_substrait.py]: Ah, thanks. In that case maybe I can remote dataset from this file completely. I'll try that.

**Comment by ianmcook** [comment]
@github-actions crossbow submit example-python-minimal-build-fedora-conda

**Pull request merged**